### PR TITLE
std: Handle ENOSYS when calling `pipe2`

### PR DESCRIPTION
**Pull request by alexcrichton**
Should help fix an accidental regression from #39386.

### CODE REVIEW

**Comment by alexcrichton** [comment]
Nominating for a beta backport as this is technically a regression from 1.17.0

File: src/libstd/sys/unix/pipe.rs [hunk 1]
```diff
@@ -35,9 +35,18 @@ pub fn anon_pipe() -> io::Result<(AnonPipe, AnonPipe)> {
     {
         weak! { fn pipe2(*mut c_int, c_int) -> c_int }
         if let Some(pipe) = pipe2.get() {
-            cvt(unsafe { pipe(fds.as_mut_ptr(), libc::O_CLOEXEC) })?;
-            return Ok((AnonPipe(FileDesc::new(fds[0])),
-                       AnonPipe(FileDesc::new(fds[1]))));
+            // Note that despite calling a glibc function here we may still get
+            // ENOSYS, see
+            // https://github.com/rust-lang/rust/pull/39386#issuecomment-306970390
+            // for more info.
```
Comment (aidanhs) [inline src/libstd/sys/unix/pipe.rs]: Can we not just copy the info here? I.e. "...may still get ENOSYS if the kernel is too old for pipe2 as glibc unconditionally defines the pipe2 symbol as a thin syscall wrapper."

Comment (alexcrichton) [inline src/libstd/sys/unix/pipe.rs]: Sure yeah!

**Review by cuviper — APPROVED** [review]
Looks good to me, thanks!

**Comment by tbu-** [comment]
This undoes what #39386 tried to do: You can save the `ENOSYS` failure of `pipe2` in a global atomic boolean (with relaxed semantics) so you won't try `pipe2` over and over again if it doesn't exist.

**Comment by cuviper** [comment]
Ah, that was before the advice to ignore `ENOSYS`.  An atomic flag seems fine, and is already used in a few other `CLOEXEC` fallbacks, but it seems frustrating that these aren't more consistent.

**Pull request merged**